### PR TITLE
Support loading license key from SSM Parameter Store

### PR DESCRIPTION
**Pull request by nc-holodakg**
Resolves https://github.com/newrelic/newrelic-lambda-extension/issues/94

### CODE REVIEW

**Review by kolanos — APPROVED** [review]

File: credentials/credentials.go [hunk 1]
```diff
@@ -77,23 +111,82 @@ func GetNewRelicLicenseKey(ctx context.Context, conf *config.Configuration) (str
...
+
+	envLicenseKey, found := os.LookupEnv(defaultSecretId)
+	if found {
+		return envLicenseKey, nil
+	}
+
+	// Check for errors while fetching from Secrets Manager or SSM Parameter Store
+	if err != nil {
+		return "", err
+	}
```
Comment (MattWhelan) [inline credentials/credentials.go:141]: This is a really weird error check. The flow control in this function is really quite confusing.

**Review by MattWhelan — CHANGES_REQUESTED** [review]
Please clarify the flow control in GetNewRelicLicenseKey. It's not clear how or when the fallbacks might run.

File: credentials/credentials.go [hunk 2]
```diff
@@ -38,7 +43,16 @@ func getLicenseKeySecretId(conf *config.Configuration) string {
 		return conf.LicenseKeySecretId
 	}
 
-	return defaultSecretId
+	return ""
```
Comment (mrickard) [inline credentials/credentials.go:46]: Are we sure that we want to return an empty string instead of the default secret ID in this case?

**Review by nc-holodakg — COMMENTED** [review]
Thanks for the feedback! A few things I'm still unsure about:
- How important it is to preserve the behavior of the fallback secret id (which doesn't seem to be documented anywhere)?
- In the edge case where both a secret id and a parameter name are configured, which should be preferred?
- Is it better to try all fallback before raising an error, or just return the first error?

File: credentials/credentials.go (hunk 1 above)
Comment (nc-holodakg) [reply credentials/credentials.go]: I was trying to handle the case where both Secrets Manager or Parameter Store were configured. Hopefully the new version is more straightforward.

**Comment by nc-wittj** [comment]
@kolanos @MattWhelan @mrickard would you be able to take look at this?

**Comment by kolanos** [comment]
@nc-holodakg Can you rebase this PR?

As for your questions:

> • How important it is to preserve the behavior of the fallback secret id (which doesn't seem to be documented anywhere)?

We want to preserve this behavior.

> • In the edge case where both a secret id and a parameter name are configured, which should be preferred?

Since secret IDs were introduced first, they should take precedence.

> • Is it better to try all fallback before raising an error, or just return the first error?

Generally, yes.

**Review by kolanos — CHANGES_REQUESTED** [review]
This generally LGTM. Would be good to have a [startup check](https://github.com/newrelic/newrelic-lambda-extension/blob/main/checks/sanity_check.go#L30) that checks if the user has configured both a secret ID and param.

File: credentials/credentials.go [hunk 3]
```diff
@@ -69,31 +81,92 @@ func IsSecretConfigured(ctx context.Context, conf *config.Configuration) bool {
...
+		util.Logln("Fetching license key from parameter name " + conf.LicenseKeyParameterName)
+		return tryLicenseKeyFromParameter(ctx, parameterName)
+	}
+
+	envLicenseKey, found := os.LookupEnv(defaultSecretId)
+	if found {
+		return envLicenseKey, nil
+	}
+
+	util.Logln("No configured license key found, attempting fallbacks")
```
Comment (MattWhelan) [inline credentials/credentials.go:123]: This should be a debug log; this is the expected flow.

File: config/config.go [hunk 4]
```diff
@@ -17,23 +17,25 @@ const (
...
+	LogLevel                string
+	SendFunctionLogs        bool
+	LogServerHost           string
 }
 
 func ConfigurationFromEnvironment() *Configuration {
 	enabledStr, extensionEnabledOverride := os.LookupEnv("NEW_RELIC_LAMBDA_EXTENSION_ENABLED")
 	licenseKey, lkOverride := os.LookupEnv("NEW_RELIC_LICENSE_KEY")
 	licenseKeySecretId, lkSecretOverride := os.LookupEnv("NEW_RELIC_LICENSE_KEY_SECRET")
+	licenseKeyParameterName, lkParameterOverride := os.LookupEnv("NEW_RELIC_LICENSE_KEY_PARAMETER")
```
Comment (MattWhelan) [inline config/config.go:38]: "PARAMETER" is really generic. It has a specific meaning here, it terms of the SSM Agent. Let's find a name that makes the context clear.

**Review by MattWhelan — COMMENTED** [review]
Let's find a more descriptive name for the new env var.

I'd also like a discussion of the Lambda execution role implications of this feature. How is this going to fail for users who have forgotten to modify their execution role? What changes are necessary? 

I'm not strongly against this; I won't block it. I'm definitely not convinced that it belongs here though. It feels like a one-off for a weird, obscure use case that we'll be saddled with maintaining. It adds a lot of complexity, for little benefit. 

The error handling is much improved though. It looks like unsurprising Go code now.

**Review by MattWhelan — DISMISSED** [review]

**Comment by nc-holodakg** [comment]
@MattWhelan @kolanos I've updated the code and environment variable naming to use "SSM Parameter" instead of "Parameter".

> Would be good to have a startup check that checks if the user has configured both a secret ID and param.

A sanity check for this is in place.

> I'd also like a discussion of the Lambda execution role implications of this feature. How is this going to fail for users who have forgotten to modify their execution role? What changes are necessary?

No changes are needed for anyone using the secret key or environment variable. If someone wants to migrate their license key to Parameter Store, they'll need to add permissions for `ssm:GetParameter`. If permissions are missing, something like this will show up in the lambda logs:
```
[NR_EXT] Failed to retrieve New Relic license key AccessDeniedException: User: arn:aws:sts::<acctid>:assumed-role/<role> is not authorized to perform: ssm:GetParameter on resource: arn:aws:ssm:us-east-1:<acctid>:parameter/<parameter_name> because no identity-based policy allows the ssm:GetParameter action
```

> It feels like a one-off for a weird, obscure use case that we'll be saddled with maintaining.

This project already supports Secrets Manager. I'm curious why you think Parameter Store is a weird edge case but Secrets Manager isn't?

> The error handling is much improved though. It looks like unsurprising Go code now.

Good to hear!

**Comment by nc-holodakg** [comment]
@kolanos @MattWhelan @mrickard Bumping this PR, would love to get this merged sometime soon

**Comment by mrickard** [comment]
@nc-holodakg This repo is now being managed by a team that concentrates on Go. I'll notify them.

**Comment by iamemilio** [comment]
Taking a look 👀

**Comment by nc-wittj** [comment]
@iamemilio did you get a chance to take a look?

**Comment by nc-wittj** [comment]
@iamemilio just wanted to check in! Did you get a chance to take a look?

**Comment by iamemilio** [comment]
Hi, sorry. I've been juggling a lot. I will test and review this by end of week. Thanks for being patient

**Comment by nc-wittj** [comment]
@iamemilio no worries!

File: config/config.go [hunk 5]
```diff
@@ -75,6 +77,8 @@ func ConfigurationFromEnvironment() *Configuration {
 		ret.LicenseKey = licenseKey
 	} else if lkSecretOverride {
 		ret.LicenseKeySecretId = licenseKeySecretId
+	} else if lkSSMParameterOverride {
```
Comment (iamemilio) [inline config/config.go:80]: I tested this in an environment with a license key in secrets manager and an SSM key, and it seems like the SSM key gets overridden if there is a key in secrets manager with the name "NEW_RELIC_LICENSE_KEY". If a customer chooses to provide a key through an environment variable explicitly, then we should always choose to use that key. Is there a way you could add that behavior to this code?

Comment (nc-holodakg) [reply config/config.go:80]: The intended behavior you described should be how this already works in `GetNewRelicLicenseKey`. I haven't tested this exact scenario myself, I'll try to reproduce what you saw.

Comment (nc-holodakg) [reply config/config.go:80]: @iamemilio Based on my testing, as long as the `NEW_RELIC_LICENSE_KEY_SSM_PARAMETER_NAME` environment variable exists, that takes precedence over a Secrets Manager secret or an SSM parameter with the `NEW_RELIC_LICENSE_KEY` name. So I believe this code is already working as you requested.

**Review by iamemilio — APPROVED** [review]

**Comment by iamemilio** [comment]
LGTM. Thanks for contributing and being patient with us :)

**Comment by nc-wittj** [comment]
@iamemilio how long does it usually take for an updated Lambda layer to be published that includes the latest extension? It looks like the most recent version for `NewRelicLambdaExtension` is 29, which was created on 2023-01-06.

```
aws lambda get-layer-version --layer-name arn:aws:lambda:us-east-1:451483290750:layer:NewRelicLambdaExtension --version-number 29
```
```json
{
    "Content": {
        "Location": "<redacted>",
        "CodeSha256": "XjFXoYnzmizYTKDDRBNLyKMCtqogwxeg6ALna9ddAh0=",
        "CodeSize": 3081073
    },
    "LayerArn": "arn:aws:lambda:us-east-1:451483290750:layer:NewRelicLambdaExtension",
    "LayerVersionArn": "arn:aws:lambda:us-east-1:451483290750:layer:NewRelicLambdaExtension:29",
    "Description": "New Relic Layer for provided (x86_64)",
    "CreatedDate": "2023-01-06T16:40:53.387+0000",
    "Version": 29,
    "CompatibleRuntimes": [
        "provided",
        "provided.al2",
        "dotnetcore3.1",
        "dotnet6"
    ],
    "LicenseInfo": "Apache-2.0",
    "CompatibleArchitectures": [
        "x86_64"
    ]
}
```

**Comment by nc-wittj** [comment]
@iamemilio any idea when this will make it into the Lambda layer? That project seems to still be on [v2.3.6](https://github.com/newrelic/newrelic-lambda-layers/blob/1321b242948d70be3e451bbede0eb61f0f12bb68/libBuild.sh#L63).

**Pull request merged**